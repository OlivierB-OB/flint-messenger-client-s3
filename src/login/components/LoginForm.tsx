import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { ILoginForm } from '../types';
import { loginUpdateForm } from '../actions/loginUpdateForm';
import { makeSubmitLogin } from '../actions/makeSubmitLogin';

interface ILoginFormDisplayProps {
  email: string;
  password: string;
  update<T extends keyof ILoginForm>(field: T, value: string): void;
  submit(): void;
}

function LoginFormDisplay({ email, password, update, submit }: ILoginFormDisplayProps) {
  return (
    <Container maxWidth="xs">
      <Box style={{ margin: '2rem 0' }}>
        <TextField
          label="Email"
          value={email}
          required={true}
          fullWidth={true}
          onChange={(event) => update('email', event.target.value)}
        />
        <TextField
          type="password"
          label="Password"
          value={password}
          fullWidth={true}
          onChange={(event) => update('password', event.target.value)}
        />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={4}>
            <Button color="primary" variant="contained" fullWidth={true} onClick={() => submit()}>
              Submit
            </Button>
          </Grid>
        </Grid>
      </Box>
    </Container>
  );
}

const mapStateToProps = ({ login }: IAppState) => ({
  ...login.form,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof ILoginForm>(field: T, value: string) => dispatch(loginUpdateForm(field, value)),
  submit: () => dispatch(makeSubmitLogin()),
});

export const LoginForm = connect(mapStateToProps, mapDispatchToProps)(LoginFormDisplay);
