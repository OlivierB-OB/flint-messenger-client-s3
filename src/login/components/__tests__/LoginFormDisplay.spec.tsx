import TextField from '@material-ui/core/TextField';
import React from 'react';
import { shallow } from 'enzyme';
import Button from '@material-ui/core/Button';
import { LoginFormDisplay } from '../LoginForm';

describe('LoginFormDisplay', () => {
  it('should display the login form', async () => {
    const props: any = {
      status: 'the status',
      email: {
        value: 'email value',
        isValid: true,
      },
      password: {
        value: 'password value',
        isValid: true,
      },
      update: 'update',
      submit: 'submit',
    };
    const component = shallow(<LoginFormDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the login form errors', async () => {
    const props: any = {
      status: 'the status',
      email: {
        value: 'email value',
        isValid: false,
        error: 'email error',
      },
      password: {
        value: 'password value',
        isValid: false,
        error: 'password error',
      },
      update: 'update',
      submit: 'submit',
    };
    const component = shallow(<LoginFormDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow editing the email value', async () => {
    const data: any = {
      status: 'the status',
      email: {
        value: 'email value',
        isValid: true,
      },
      password: {
        value: 'password value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<LoginFormDisplay {...data} />);
    component
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('email', 'new value');
  });

  it('should allow editing the password value', async () => {
    const data: any = {
      status: 'the status',
      email: {
        value: 'email value',
        isValid: true,
      },
      password: {
        value: 'password value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<LoginFormDisplay {...data} />);
    component
      .find(TextField)
      .at(1)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('password', 'new value');
  });

  it('should allow submitting', async () => {
    const props: any = {
      status: 'the status',
      email: {
        value: 'email value',
        isValid: true,
      },
      password: {
        value: 'password value',
        isValid: true,
      },
      submit: jest.fn(),
    };
    const component = shallow(<LoginFormDisplay {...props} />);
    expect(component.find(Button).at(0).prop('type')).toEqual('submit');
    component.find('form').simulate('submit', { preventDefault: () => null });
    expect(props.submit).toHaveBeenCalled();
  });
});
