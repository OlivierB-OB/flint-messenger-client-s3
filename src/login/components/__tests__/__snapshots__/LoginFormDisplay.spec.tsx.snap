// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFormDisplay should display the login form 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="xs"
>
  <Styled(MuiBox)
    style={
      Object {
        "margin": "2rem 0",
      }
    }
  >
    <Alert
      status="the status"
    />
  </Styled(MuiBox)>
  <form
    onSubmit={[Function]}
  >
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        label="Email"
        onChange={[Function]}
        required={true}
        value="email value"
      />
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        label="Password"
        onChange={[Function]}
        required={true}
        type="password"
        value="password value"
      />
    </Styled(MuiBox)>
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="flex-end"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            fullWidth={true}
            type="submit"
            variant="contained"
          >
            Submit
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
  </form>
</WithStyles(ForwardRef(Container))>
`;

exports[`LoginFormDisplay should display the login form errors 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="xs"
>
  <Styled(MuiBox)
    style={
      Object {
        "margin": "2rem 0",
      }
    }
  >
    <Alert
      status="the status"
    />
  </Styled(MuiBox)>
  <form
    onSubmit={[Function]}
  >
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(TextField))
        error={true}
        fullWidth={true}
        helperText="email error"
        label="Email"
        onChange={[Function]}
        required={true}
        value="email value"
      />
      <WithStyles(ForwardRef(TextField))
        error={true}
        fullWidth={true}
        helperText="password error"
        label="Password"
        onChange={[Function]}
        required={true}
        type="password"
        value="password value"
      />
    </Styled(MuiBox)>
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="flex-end"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            fullWidth={true}
            type="submit"
            variant="contained"
          >
            Submit
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
  </form>
</WithStyles(ForwardRef(Container))>
`;
